Fix use of deprecated `check_no_isolation` in posix fs shims

atsmtat opened the pull request:
Update posix fs shims to use new API `reject_in_isolation`, which
allows rejection with error code instead of always forcing abort.
Error code chosen for each op is the most appropriate one from the
list in corresponding syscall's manual.

Updated helper APIs to not use quotes (\`) around input name while
preparing the message. This allows callers to pass multi-word string
like -- "\`read\` from stdin".

Cc https://github.com/rust-lang/miri/issues/1034

RalfJung commented on src/shims/posix/fs.rs:
```diff
@@ -827,7 +847,15 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        if let IsolatedOp::Reject(reject_with) = this.machine.isolated_op {
+            this.reject_in_isolation("`stat`", reject_with)?;
+            // macos stat never sets "EPERM". Set error code "ENOENT".
+            this.set_last_error_from_io_error(ErrorKind::NotFound)?;
```
Is there no more accurate error code? What do macOS sandboxes typically do?

atsmtat replied:
This variant of `stat` works on a path name and doesn't need read, write or execute permission. `ENOENT` corresponds to file does not exist. Another relevant code is `EACCES` -- permission denied for a component of the path prefix. I'm not sure about macOS sandboxes. I guess `EACCES` sounds closer to `EPERM`.

RalfJung replied:
yeah EACCESS would be better I think

RalfJung commented on src/shims/posix/fs.rs:
```diff
@@ -874,7 +916,14 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        if let IsolatedOp::Reject(reject_with) = this.machine.isolated_op {
+            this.reject_in_isolation("`statx`", reject_with)?;
+            // statx never sets "EPERM". Set error code "ENOENT".
+            this.set_last_error_from_io_error(ErrorKind::NotFound)?;
```
It looks like there is also `EACCES` -- would that be more accurate?

atsmtat replied:
Since `statx` supports both path name and fd, we'll have to set error code based on the input (pathname vs. fd) to be more accurate. So I moved this logic after inputs are resolved.

RalfJung replied:
> Since statx supports both path name and fd, we'll have to set error code based on the input (pathname vs. fd) to be more accurate.

I don't quite understand this comment...

RalfJung replied:
Oh I saw the new code, makes sense.

RalfJung commented:
Could you also add a test that exercises these functions? Something like `tests/run-pass/current_dir_with_isolation.rs` (but it doesn't need to do that loop, calling the function a single time suffices).

RalfJung commented:
@atsmtat it's been a while since we heard from you, what is the status of this PR?

I'm afraid there will probably be a lot of conflicts with https://github.com/rust-lang/miri/pull/1851.

atsmtat commented:
> @atsmtat it's been a while since we heard from you, what is the status of this PR?
> 
> I'm afraid there will probably be a lot of conflicts with #1851.

Sorry for the late reply. I got busy. I addressed your comments in a new commit I just pushed.

I can wait for your PR to merge and then rebase and resolve conflicts.

atsmtat commented:
> > @atsmtat it's been a while since we heard from you, what is the status of this PR?
> > I'm afraid there will probably be a lot of conflicts with #1851.
> 
> Sorry for the late reply. I got busy. I addressed your comments in a new commit I just pushed.
> 
> I can wait for your PR to merge and then rebase and resolve conflicts.

Just noticed that your PR is merged. I rebased my branch and resolved conflicts, wasn't too bad 🙂

RalfJung commented on src/shims/posix/fs.rs:
```diff
@@ -852,7 +888,13 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+            return this.handle_not_found();
+        }
 
         let fd = this.read_scalar(fd_op)?.to_i32()?;
```
One thing I noticed in my last PR (and fixed for the existing code): we should make sure that the argument checks (`read_scalar`, `read_pointer`, and so on) are all *before* the isolation check. That way, extremely incorrect arguments (like, using the wrong function prototype, or passing uninit data) are still detected.

This applies to all the functions you touched in this file, I think.

atsmtat replied:
It makes sense. I'll move these checks right before the system call ops in shims. Will update the PR soon.

RalfJung commented on src/shims/posix/fs.rs:
```diff
@@ -924,6 +981,22 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                this.eval_libc("EACCES")?
+            } else {
+                // `dirfd` is set to target file, and `path` is
+                // empty. `EACCES` would violate the spec.
```
Why is path empty? It could be relative, no?

atsmtat replied:
`if` block handles the case of `path` being relative. In that case, dirfd has to be set to `AT_FDCWD`.

RalfJung replied:
```suggestion
                // empty (or we would have hit the `throw_unsup_format` above). `EACCES` would violate the spec.
                assert!(empty_path_flag);
```
Just making sure we notice if the code above ever changes and someone forgets to update this code here.

atsmtat replied:
Yeah, I was also not completely happy repeating the valid input checks. One more option is, I can combine two if statements into one, avoiding code duplication.

RalfJung replied:
No I think I like the code organization here. Just please add the `assert!`.

atsmtat replied:
Alright, done.

RalfJung commented on src/shims/posix/fs.rs:
```diff
@@ -827,9 +847,25 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+            None => return Ok(-1),
+        };
+
+        this.macos_stat_write_buf(metadata, buf_op)
```
`macos_stat_write_buf` is unused now and can be removed, right?

atsmtat replied:
I guess you meant `macos_stat_or_lstat`. Right, I forgot to delete it and I  guess rustc didn't complain because it's a trait function? Thanks for pointing out.

RalfJung replied:
Ah yes I dd, sorry for the confusion.

RalfJung commented:
Awesome, thanks a lot. :)
@bors r+

bors commented:
:pushpin: Commit 20d0f2ee2607a1eecf095223250ca07307b9dbb0 has been approved by `RalfJung`

<!-- @bors r=RalfJung 20d0f2ee2607a1eecf095223250ca07307b9dbb0 -->
<!-- homu: {"type":"Approved","sha":"20d0f2ee2607a1eecf095223250ca07307b9dbb0","approver":"RalfJung"} -->

bors commented:
:hourglass: Testing commit 20d0f2ee2607a1eecf095223250ca07307b9dbb0 with merge 1677946d74e411b497cf06e69749e7e1b15afa09...
<!-- homu: {"type":"BuildStarted","head_sha":"20d0f2ee2607a1eecf095223250ca07307b9dbb0","merge_sha":"1677946d74e411b497cf06e69749e7e1b15afa09"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/miri/runs/3154236385)
Approved by: RalfJung
Pushing 1677946d74e411b497cf06e69749e7e1b15afa09 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"RalfJung","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/miri/runs/3154236385"},"merge_sha":"1677946d74e411b497cf06e69749e7e1b15afa09"} -->
